[ROCm] Update CI dockers to ROCm release 3.1.1

iotamudelta opened the pull request:
Request to update ROCm CI dockers to release 3.1

Changes required to the PyTorch source base attached:
* switch to the fast path for the Caffe2 ReLU operator
* switch to the new hipMemcpyWithStream(stream) API to replace hipMemcpyAsync(stream) && hipStreamSynchronize(stream) paradigm in an optimized fashion
* disable two regressed unit tests

iotamudelta committed:
Change Caffe2 ReLU operator to fast path

iotamudelta committed:
Introduce hipMemcpyWithStream()

Optimized API for the hipMemcpyAsync(stream) &&
hipStreamSynchronize(stream) paradigm.

iotamudelta committed:
Disable two tests that have regressed with ROCm 3.1

ezyang commented:
Not passing CI

```
04:28:31 In file included from /var/lib/jenkins/workspace/aten/src/THH/THHTensorMathReduce.cuh:9:
04:28:31 /var/lib/jenkins/workspace/aten/src/THH/THHReduceAll.cuh:325:17: error: use of undeclared identifier 'hipMemcpyWithStream'
04:28:31     THCudaCheck(hipMemcpyWithStream(out,
04:28:31                 ^
```

iotamudelta commented:
@ezyang this API requires CI to be on ROCm 3.1 - hence the request to update in lockstep with this PR. :-)

ezyang commented:
OK. Who is going to do the docker update?

iotamudelta commented:
@ezyang would love to do it but don't think I have the required permissions. Either you or @bddppq would need to lend us a hand. I can put the base docker images up if that helps.

ezyang commented:
Unfortunately, @bddppq isn't working at FB anymore :( I'll see if I can find someone to take point.

ezyang commented:
cc @seemethere

iotamudelta committed:
Port PyTorch Ubuntu docker over to circleci setup.

Update numpy version Caffe2 to be again in sync with PyTorch.

ezyang commented on .circleci/docker/common/install_rocm.sh:
```diff
@@ -0,0 +1,75 @@
+                   rocthrust \
+                   rccl \
+                   rocprofiler-dev \
+                   roctracer-dev
```
We don't have to fix this in this diff, but similar to the other installation scripts, you should clear the apt cache when you're done to reduce the size of the docker image.

ezyang commented:
@iotamudelta OK, you know how to trigger a build of the docker images, right? You'll have to push your branch to pytorch main repo, rather than just having it in your fork.

pruthvistony commented on .circleci/docker/common/install_rocm.sh:
```diff
@@ -0,0 +1,75 @@
+    # Need the libc++1 and libc++abi1 libraries to allow torch._C to load at runtime
+    apt-get install libc++1
+    apt-get install libc++abi1
+
```
Can we add libsndfile1 for ubuntu and libsndfile for centos so that it can removed from .jenkins/pytorch/test.sh.
Tried to do that in PR#34465, but its not required now.

iotamudelta replied:
I'll add that to the follow-up PR which will also do the apt-cache cleanup.

iotamudelta committed:
Remove skips - regression fixed in ROCm 3.1.1

xw285cornell commented:
@iotamudelta were you able to build a new docker image via our jenkins, and run this PR against the new docker image?

xw285cornell reviewed (changes requested):
The ubuntu image build has failed with the following message: https://ci.pytorch.org/jenkins/job/caffe2-docker/job/py3.6-clang7-rocmdeb-ubuntu16.04/99/console

09:41:31 [91mE: The method driver /usr/lib/apt/methods/https could not be found.
09:41:31 [0mThe command '/bin/sh -c if [ -n "${CLANG_VERSION}" ]; then bash ./install_clang.sh; fi' returned a non-zero code: 100

You probably need `apt-get install apt-transport-https ca-certificates`.

iotamudelta commented:
@xw285cornell thanks! After talking to @ezyang I'm using iotamudelta/update_rocm3.1 for staging and have added that. Will update anything here once that works.

xw285cornell commented:
I see, sounds good. I just caught up with the slack channel -- iirc we should build a new docker image and get a tag, and test this PR against the new docker to make sure everything works. And then we can merge the docker image update with this PR I think.

iotamudelta committed:
Put missing bit in

iotamudelta commented:
Docker image build for CentOS and Ubuntu successful: https://ci.pytorch.org/jenkins/job/caffe2-docker-trigger/369/console (based on iotamudelta/update_rocm3.1 branch)

xw285cornell commented:
Kicking off a run with 369 docker here: 

https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py3.6-devtoolset7-rocmrpm-centos7-trigger-test/17345/

https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/58975/

iotamudelta commented:
Oh, OK - I already kicked: https://ci.pytorch.org/jenkins/job/pytorch-pull-request/97002/ off for the PyTorch ones as well.

Concerning py2, I don't think we are supporting that any longer?

xw285cornell commented:
Sounds good, let's see how it goes. 

But wait, I think we need to upgrade the drivers on the CI machines first? Will the jobs fail? 

Regarding py2, yeah I kicked off the run with the wrong config, just posted it there since I already kicked if off. If it failed with py2 failures it's ok (just there to detect if there is any flakeness).

iotamudelta commented:
Yes, I would assume that - this should be valid for compilation though. Let's get the hosts updated next.

iotamudelta commented:
Looks like both https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py3.6-clang7-rocmdeb-ubuntu16.04-build/27488/console and https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py3.6-devtoolset7-rocmrpm-centos7-build/17330/console were successfully built with the new docker images! :-)

iotamudelta commented:
@ezyang @xw285cornell https://ci.pytorch.org/jenkins/job/pytorch-pull-request/97002/ looks like everything is green, including tests - even on the old host kernel. So once the host kernels are updated, we should be ready to land?

xw285cornell commented:
@iotamudelta we'll have to coordinate this diff with internal CI, as it introduces new API. I'll try to find some time to do that. If possible -- can we break this diff into two, one upgrade the docker image, and one introduce the new API? If so, we can land the docker upgrade sooner. 

Also I didn't see caffe2 use this new API, would you include there as well?

iotamudelta commented:
How about putting an version guard in - this seems to be a natural choice since we are anyways ifdefing?

Concerning Caffe2: I've not seen the hipMemcpyAsync() && hipStreamSynchronize() paradigm as plainly visible there.

xw285cornell commented:
Yes, version guard will be great. Any chance we can have caffe2 code get benefit from this new API?

iotamudelta committed:
Update tag to the ROCm 3.1.1 image

ezyang commented:
@iotamudelta I can merge this whenever, let me know when.

ezyang commented:
(preferably on slack, for timeliness)

iotamudelta committed:
ifdef use of newer APIs to ROCm 3.1+

Try to set the HIP_VERSION in accordance w/ the actual HIP definition

iotamudelta committed:
Fix typo.

iotamudelta commented:
Looks like ROCm CI targets vanished.

xw285cornell commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -174,13 +174,17 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
   } else {
+#if HIP_VERSION >= 310
+    AT_CUDA_CHECK(hipMemcpyWithStream(dst, src, nbytes, kind, stream));
+#else
```
Just want to make sure I understand this new API -- hipMemcpyWithStream waits for the entire copy to finish on a stream, is that right? So if I have multi-threads launching the copy, the copy can be in parallel -- but the API will hold each thread until it finishes?

iotamudelta replied:
@xw285cornell concerning hipMemcpyWithStream, it directly replaces, optimizes, and behaves as if you'd write hipMemcpyAsync(stream) && hipStreamSynchronize(stream). So serialization is local to the stream (of course null stream has special properties here). So your statement is correct, multiple streams can be launched to and copy in parallel, but the API will be synchronous in that stream.

xw285cornell replied:
I see sounds good. Just curious, since you mentioned this API will bring perf gains, just curious, what kind of optimizations happen under the hood for this optimization? 

The use case I was thinking about was unpinned memory, but in a different scenario. If you stay within pytorch, all cpu memory should be pinned.

xw285cornell commented on torch/lib/c10d/CMakeLists.txt:
```diff
@@ -11,7 +11,8 @@ if(USE_CUDA)
   add_definitions(-D__HIP_PLATFORM_HCC__=1)
-  add_definitions(-DHIP_VERSION=${HIP_VERSION_MAJOR})
+  # same definition as in hip/hip_version.h
+  add_definitions(-DHIP_VERSION=(${HIP_VERSION_MAJOR}*100+${HIP_VERSION_MINOR}))
```
Can this be defined in some Pytorch files (I recall there is a place where we define miopen/cudnn versions). Internally we don't use Makefile, so this will still break internal ci.

iotamudelta replied:
I'm just including the hip_version.h file in applicable files and removed the HIP_VERSION setting in make systems - I don't think that was a good strategy since it overrides the HIP included ones with a lower number (it's HIP_MAJOR_VERSION vs HIP_MAJOR_VERSION*100+HIP_MINOR_VERSION)

xw285cornell replied:
Why not include hip_version.h in a higher-level file, so that we can use it anywhere, without explicitly include it? 

Also -- you mentioned this file is introduced in ROCm 3. Can we add some guard to that include?

xw285cornell commented on .circleci/verbatim-sources/workflows-ecr-gc.yml:
```diff
@@ -14,7 +14,7 @@
             name: ecr_gc_job_for_caffe2
             project: caffe2
-            tags_to_keep: "348,345,336,325,324,315,306,301,287,283,276,273,266,253,248,238,230,213"
+            tags_to_keep: "369,348,345,336,325,324,315,306,301,287,283,276,273,266,253,248,238,230,213"
```
I'm not an expert in CI, but I think there are some other places where you need to specify to use this tag (this one probably just prevent the image to be purged). And you need to run a generate script. (iirc https://github.com/pytorch/pytorch/blob/master/.circleci/cimodel/data/caffe2_build_definitions.py#L15)

iotamudelta replied:
Yeah, @ezyang is helping with all of that! :-)

iotamudelta committed:
Directly include hip/hip_version.h

Remove setting of HIP_VERSION in make files - this should always be
handled with the include of the version file.

iotamudelta committed:
Fix formatting issue

xw285cornell commented:
I started internal upgrade to 3.1, and I'm seeing some issues with relu_op which is fixed in this patch. Going to stack this one to get the build work.

xw285cornell commented:
Created FBA172-176 to track various bugs I'm facing

xw285cornell commented:
The internal CI is happy, however I think this PR needs to be updated to get the new docker.

iotamudelta committed:
Update to the new image tag

xw285cornell commented:
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/98978/
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/98979/
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/98980/
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/98981/
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/98982/

iotamudelta committed:
Merge remote-tracking branch 'rocm_upstream/upstream' into update_rocm3.1

iotamudelta committed:
Disable test introduced in
https://github.com/pytorch/pytorch/pull/35219

which also breaks on ROCm 2.9

iotamudelta committed:
Disable test fallout

iotamudelta committed:
Disable test fallout

iotamudelta committed:
Also import.

iotamudelta committed:
Regenerate config.

ezyang commented:
```

18:54:20 In file included from /var/lib/jenkins/workspace/aten/src/THH/THHTensorMathScan.hip:9:
18:54:20 In file included from /var/lib/jenkins/workspace/aten/src/THH/THHTensorMathReduce.cuh:9:
18:54:20 /var/lib/jenkins/workspace/aten/src/THH/THHReduceAll.cuh:17:10: fatal error: 'hip/hip_version.h' file not found
18:54:20 #include <hip/hip_version.h>
18:54:20          ^~~~~~~~~~~~~~~~~~~
```

iotamudelta commented:
@ezyang are we still on the old docker image?

xw285cornell commented:
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/99252/
https://ci.pytorch.org/jenkins/job/pytorch-pull-request/99253/

iotamudelta committed:
Fix annotation.

iotamudelta committed:
Merge remote-tracking branch 'upstream/master' into update_rocm3.1

xw285cornell commented:
I'm rebuilding with onnx docker here: https://ci.pytorch.org/jenkins/job/caffe2-docker-trigger/373/

ezyang commented:
Landing.

peterjc123 commented:
Looks like this PR is breaking the Windows CI.
```
C:/Users/circleci/project/aten/src/ATen/native/cuda/CUDAScalar.cu(16): error: "#" not expected here

C:/Users/circleci/project/aten/src/ATen/native/cuda/CUDAScalar.cu(16): error: expected an expression
```

iotamudelta committed:
Merge remote-tracking branch 'rocm_upstream/upstream' into update_rocm3.1

iotamudelta committed:
Try to pacify Windows.

iotamudelta commented:
Changed CUDAScalar for Windows. Any fallout should be visible in here.

iotamudelta committed:
Update tag to 373

iotamudelta committed:
Skip fully.

iotamudelta committed:
Merge branch 'master' into update_rocm3.1

iotamudelta committed:
Merge remote-tracking branch 'upstream/master' into update_rocm3.1

iotamudelta committed:
regenerate config

iotamudelta committed:
Add 373

iotamudelta committed:
Regenerate

iotamudelta committed:
Update and regen

iotamudelta committed:
Remove tab
